README.md
# TDAstats: topological data analysis in R <img src="man/figures/HexTDA.png" align="right" height="175" width="151"/>

[![Travis-CI Build Status](https://travis-ci.org/rrrlw/TDAstats.svg?branch=master)](https://travis-ci.org/rrrlw/TDAstats)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/rrrlw/TDAstats?branch=master&svg=true)](https://ci.appveyor.com/project/rrrlw/TDAstats)
[![Coverage Status](https://img.shields.io/codecov/c/github/rrrlw/TDAstats/master.svg)](https://codecov.io/github/rrrlw/TDAstats?branch=master)

[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![CRAN version](http://www.r-pkg.org/badges/version/TDAstats)](https://CRAN.R-project.org/package=TDAstats)
[![CRAN Downloads](http://cranlogs.r-pkg.org/badges/grand-total/TDAstats)](https://CRAN.R-project.org/package=TDAstats)

[![JOSS DOI](http://joss.theoj.org/papers/10.21105/joss.00860/status.svg)](https://doi.org/10.21105/joss.00860)
[![Zenodo DOI](https://zenodo.org/badge/130141540.svg)](https://zenodo.org/badge/latestdoi/130141540)

## Overview

TDAstats is an R pipeline for computing persistent homology in topological data analysis.

## Installation

To install TDAstats, run the following R code:
```r
# install from CRAN
install.packages("TDAstats")

# install development version from GitHub
devtools::install_github("rrrlw/TDAstats")

# install development version with vignettes/tutorials
devtools::install_github("rrrlw/TDAstats", build_vignettes = TRUE)
```

## Sample code

The following sample code creates two synthetic datasets, and calculates and visualizes their persistent homology to showcase the use of TDAstats.

```r
# load TDAstats
library("TDAstats")

# load sample datasets
data("unif2d")
data("circle2d")

# calculate persistent homology for both datasets
unif.phom <- calculate_homology(unif2d, dim = 1)
circ.phom <- calculate_homology(circle2d, dim = 1)

# visualize first dataset as persistence diagram
plot_persist(unif.phom)

# visualize second dataset as topological barcode
plot_barcode(circ.phom)
```

A more detailed tutorial can be found in the package vignettes or at [this Gist](https://gist.github.com/rrrlw/2fd22a834a883cb66454b1dabab9fdcb).

## Functionality

TDAstats has 3 primary goals:

1.  *Calculation of persistent homology*: the C++
[Ripser](https://github.com/Ripser/ripser)
project is a lightweight library for calculating persistent homology
that outpaces all of its competitors. Given the importance of computational
efficiency, TDAstats naturally uses Ripser behind the scenes for homology
calculations, using the Rcpp package to integrate the C++ code into an R
pipeline (Ripser for R).

2.  *Statistical inference of persistent homology*: persistent homology can be
used in hypothesis testing to compare the topological structure of two point
clouds. TDAstats uses a permutation test in conjunction with the Wasserstein
metric for nonparametric statistical inference.

3.  *Visualization of persistent homology*: persistent homology is visualized
using two types of plots - persistence diagrams and topological barcodes.
TDAstats provides implementations of both plot types using the
[ggplot2](https://github.com/tidyverse/ggplot2)
framework. Having ggplot2 underlying the plots confers many advantages to the
user, including generation of publication-quality plots and customization using
the ggplot object returned by TDAstats.

## Contribute

To contribute to TDAstats, you can create issues for any bugs/suggestions on the [issues page](https://github.com/rrrlw/TDAstats/issues). You can also fork the TDAstats repository and create pull requests to add features you think will be useful for users.

## Citation

If you use TDAstats, please consider citing the following (based on use):
* **General use of TDAstats**: Wadhwa RR, Williamson DFK, Dhawan A, Scott JG. TDAstats: R pipeline for computing persistent homology in topological data analysis. *Journal of Open Source Software*. 2018; 3(28): 860. doi: [10.21105/joss.00860](https://doi.org/10.21105/joss.00860)
* **TDAstats to calculate persistent homology (Ripser)**: Bauer U. Ripser: Efficient computation of Vietoris-Rips persistence barcodes. 2019; *arXiv*: 1908.02518.
* **TDAstats to perform statistical test**: Robinson A, Turner K. Hypothesis testing for topological data analysis. *J Appl Comput Topol*. 2017; 1: 241.

## Real-world applications, use cases, and mentions

* Pinto GVF. Motivic constructions on graphs and networks with stability results. Doctoral Thesis: Universidade Estadual Paulista Rio Claro & Ohio State University. 2020. Link to [thesis](http://hdl.handle.net/11449/192494).
* Gommel M. A Machine Learning Exploration of Topological Data Analysis Applied to Low and High Dimensional fMRI Data. Doctoral Thesis: University of Iowa. 2019. doi: [10.17077/etd.005247](https://doi.org/10.17077/etd.005247). Link to [thesis](https://iro.uiowa.edu/discovery/fulldisplay/alma9983779398602771/01IOWA_INST:ResearchRepository?tags=scholar).
* Mémoli F, Singhal K. A Primer on Persistent Homology of Finite Metric Spaces. Bulletin of Mathematical Biology. 2019; 81(7): 2074. Links to [paper](https://link.springer.com/article/10.1007/s11538-019-00614-z) and [preprint](https://arxiv.org/abs/1905.13400)
* Srinivasan R, Chander A. Understanding Bias in Datasets using Topological Data Analysis. Fujitsu Laboratories of America. 2019. [Link](http://ceur-ws.org/Vol-2419/paper_9.pdf)
* Kough D, Neuzil M, Simpson C, Glover R. Analyzing State of the Union Addresses using Topology. University of St. Thomas. 2019. [Link](https://www.stthomas.edu/media/collegeofartsandsciences/mathematics/pdf/camsummer2019/CAMReport2019TDANeuzilKoughSimpson.pdf)
* Rickert J. A Mathematician's Perspective on Topological Data Analysis and R. 2018. [Link](https://rviews.rstudio.com/2018/11/14/a-mathematician-s-perspective-on-topological-data-analysis-and-r/)
* [Blog post on Data Management](https://www.kaisataipale.net/blog/2019/02/22/data-management/)
* [Analyzing finance data](https://github.com/kaitai/Example-with-TDAstats)
* [R package for visualizing persistent homology](https://github.com/rrrlw/ggtda)


NEWS.md
# TDAstats 0.5.0

## Changes

* `ripserr` has been added as a dependency
* All C++ code (and associated `Rcpp` code and dependencies) has been removed
from TDAstats and ported to `ripserr` (v0.1.1 on CRAN)
* `calculate_homology` is now deprecated in favor of `ripserr::vietoris_rips`
and `ripserr::cubical`

# TDAstats 0.4.1

## Bug Fixes

* fixed potential error with updates in r-devel (`matrix` inherits from `array`)

## Changes

* `calculate_homology` now accepts any variable storing a distance matrix as long as it can be coerced to a matrix

## Additions

* added option for `calculate_homology` to return a data frame instead of a matrix
* more code coverage
* added support for flat persistence diagrams (`flat = TRUE` in `plot_persist`)

# TDAstats 0.4.0

## Bug Fixes

* previous vignettes did not have citations and references section working properly; this has been fixed

## Changes

* added `standardize` parameter to `calculate_homology` and functionality to standardize point cloud size
* added `limit.num` parameter to `phom.dist` and functionality to consider only "significant" topological features

## Additions

* code coverage has been increased
* added support for distance matrix format for calculate_homology
* vignette has been added showing use of new distance matrix format
* `phom.dist` function to compare persistent homology of two point clouds/matrices

# TDAstats 0.3.0

## Bug fixes

* fixed bug in `plot_persist` where setting a shorter axis limit would cause the reference diagonal to disappear (thank you, @corybrunson)

## Changes

* changing size of `plot_persist` output figures still preserves one-to-one length ratio of horizontal and vertical axes (fixed coordinate system; thank you, @corybrunson)

# TDAstats 0.2.0

## Bug fixes

* fixed bug in `permutation_test` function that resulted in NAs in output
* fixed naming clash with `index_t` in C++ code; should make TDAstats compatible with Solaris systems

## Changes

* added `dim` parameter to `permutation_test` function to allow users to select maximum dimension for homology comparison

## Additions

* added automated testing using `testthat` package
* 4 sample datasets for users to learn/test TDAstats (and probably to follow future vignettes): unif2d, unif3d, circle2d, sphere3d
* added 2 vignettes introducing features of TDAstats to user through text and a case study

# TDAstats 0.1.0

* This is the first release of TDAstats.


R/inference.R
#####WASSERSTEIN CALCULATION#####
# convenience function using `wass_mat_calc` as main (`wass_workhorse` indirectly)
# as main workhorse
# ellipsis includes dim, format, threshold for ripserr::vietoris_rips
wass_cloud_calc <- function(pts1, pts2, pow.val = 1, ...) {
  # make sure pts1 and pts2 (matrices) have same # of cols
  if (ncol(pts1) != ncol(pts2)) {
    stop("Something wrong in code here; invalid arguments (unequal number of columns in point clouds.")
  }

  # calculate persistent homology
  phom1 <- ripserr::vietoris_rips(pts1, return_format = "mat", ...)
  phom2 <- ripserr::vietoris_rips(pts2, return_format = "mat", ...)
  
  # check if dim is present as parameter within ...
  dim <- 1 # default
  list.version <- list(...)
  if (length(list.version) > 0) {
    if (!is.null(list.version$dim)) {
      dim <- as.integer(list.version$dim)
    }
  }

  # return Wasserstein metric
  return(wass_mat_calc(phom1, phom2, pow.val = pow.val, dim = dim))
}

# convenience function using `wass_workhorse` as main workhorse
# mat1 and mat2 are feature matrices for which we want to calculate the
# Wasserstein metric
wass_mat_calc <- function(mat1, mat2, pow.val = 1, dim = 1) {
  # only want to calculate Wasserstein for overlapping dimensions
  min.dim <- 0
  max.dim <- dim
  
  # make workhorse function do all the work
  wass.vals <- vapply(X = min.dim:max.dim,
                      FUN.VALUE = numeric(1),
                      FUN = function(curr.dim) {
                        # subset matrices for current dimension only
                        curr.mat1 <- matrix(mat1[mat1[, 1] == curr.dim, ], ncol = 3)
                        curr.mat2 <- matrix(mat2[mat2[, 1] == curr.dim, ], ncol = 3)

                        #print(dim(curr.mat1))
                        #if (nrow(curr.mat1) == 0 |
                        #    nrow(curr.mat2) == 0) {
                        #  stop("Calculating Wasserstein for invalid dimension")
                        #}
                        # if there's no features for a given dimension in current permutation
                        # add a 
                        if (nrow(curr.mat1) == 0) {
                          curr.mat1 <- matrix(c(curr.dim, 0, 0), ncol = 3)
                        }
                        if (nrow(curr.mat2) == 0) {
                          curr.mat2 <- matrix(c(curr.dim, 0, 0), ncol = 3)
                        }

                        # calculate feature lengths from feature matrices
                        curr.feat1 <- curr.mat1[, 3] - curr.mat1[, 2]
                        curr.feat2 <- curr.mat2[, 3] - curr.mat2[, 2]

                        # return Wasserstein metric for current dimension
                        this.ans <- wass_workhorse(curr.feat1, curr.feat2, pow.val)
                        if (is.na(this.ans)) stop("Error: found an NA somewhere1")
                        #print(paste("This ans:", this.ans))
                        return(this.ans)
                      })

  # return vector of Wasserstein metrics (one per dimension)
  names(wass.vals) <- as.character(min.dim:max.dim)
  return(wass.vals)
}

# vec1 and vec2 are numeric vectors containing the duration (death - birth) of
#   all the features (of the same dimension) of their respective calculated
#   persistent homologies
wass_workhorse <- function(vec1, vec2, pow.val = 1) {
  # make sure vec1 and vec2 are of the same length
  if (length(vec1) < length(vec2)) {
    vec1 <- c(vec1, rep(0, length(vec2) - length(vec1)))
  } else if (length(vec2) < length(vec1)) {
    vec2 <- c(vec2, rep(0, length(vec1) - length(vec2)))
  }

  # make sure vec1 and vec2 are now of equal length
  # user should never see this
  if (length(vec1) != length(vec2)) stop("Something wrong with length-equaling code above (inference.R; wass_calc).")
  if (length(vec1) == 0) stop("Error with feature calculating code.")
  
  # sort both
  vec1 <- sort(vec1, decreasing = TRUE)
  vec2 <- sort(vec2, decreasing = TRUE)

  # make vector of absolute value of differences
  diff.vec <- abs(vec1 - vec2)

  # return answer (incorporating power value parameter)
  ans <- sum(diff.vec ^ pow.val)
  if (is.na(ans)) {
    stop("Error: found an NA somewhere2")
  }
  return(ans)
}

# get top n features from each dimension
phom.topn <- function(phom, n) {
  # all the dimensions
  dims <- unique(phom[, 1])
  phom.adjust <- matrix(NA, nrow = 0, ncol = 3)
  
  # go through all dimensions
  for (curr.dim in dims) {
    # get top n rows from this dimension
    curr.phom <- phom[phom[, 1] == curr.dim, ]
    names(curr.phom) <- NULL
    curr.phom <- matrix(curr.phom, ncol = 3)
    
    # order
    curr.phom <- cbind(curr.phom,
                       curr.phom[, 3] - curr.phom[, 2])
    curr.phom <- curr.phom[order(curr.phom[, 4], decreasing = TRUE), ]
    curr.phom <- matrix(curr.phom, ncol = 4)
    
    # combine phom.adjust - guaranteed at least one row
    phom.adjust <- rbind(phom.adjust,
                         curr.phom[1:min(nrow(curr.phom), n), 1:3])
  }
  
  return(phom.adjust)
}

#####BOOTSTRAPPING#####
#' Identify Significant Features in Persistent Homology
#' 
#' An empirical method (bootstrap) to differentiate between features that
#' constitute signal versus noise based on the magnitude of their
#' persistence relative to one another. Note: you must have at
#' least 5 features of a given dimension to use this function.
#' 
#' @param features 3xn data frame of features; the first column must be
#'   dimension, the second birth, and the third death
#' @param dim dimension of features of interest
#' @param reps  number of replicates
#' @param cutoff  percentile cutoff past which features are considered
#'   significant
#' @examples
#' # get dataset (noisy circle) and calculate persistent homology
#' angles <- runif(100, 0, 2 * pi)
#' x <- cos(angles) + rnorm(100, mean = 0, sd = 0.1)
#' y <- sin(angles) + rnorm(100, mean = 0, sd = 0.1)
#' annulus <- cbind(x, y)
#' phom <- ripserr::vietoris_rips(annulus, return_format = "mat")
#' 
#' # find threshold of significance
#' # expecting 1 significant feature of dimension 1 (Betti-1 = 1 for annulus)
#' thresh <- id_significant(features = as.data.frame(phom),
#'                          dim = 1,
#'                          reps = 500,
#'                          cutoff = 0.975)
#'
#' # generate flat persistence diagram
#' # every feature higher than `thresh` is significant
#' plot_persist(phom, flat = TRUE)
#' @export
id_significant <- function(features, dim = 1,
                           reps = 100, cutoff = 0.975) {
  # subset only features of dimension of interest
  colnames(features) <- c("dimension", "birth", "death")
  features <- features[features[, 1] == dim, ]
  
  # make sure conditions are met
  if (nrow(features) < 3) {
    stop(paste("There are too few (< 3) features of the",
               "desired dimension in the feature matrix."))
  }
  
  # calculate persistence
  features$persist <- features$death - features$birth
  
  # do bootstrap
  ans <- numeric(reps)
  for (i in 1:reps) {
    curr_sample <- sample(x = features$persist,
                          size = nrow(features),
                          replace = TRUE)
    ans[i] <- mean(curr_sample)
  }
  
  # return threshold above which features count as "significant"
  stats::quantile(ans, cutoff, names = FALSE)
}

#####DISTANCE BETWEEN PERSISTENT HOMOLOGY#####
# calculate distance between two persistent homology matrices (filled with features)
# inspired from Wasserstein/EMD but not the same (no intrinsic ordering of features, etc.)
#' Calculate Distance between Homology Matrices
#' 
#' Calculates the distance between two matrices containing persistent homology
#' features, usually as returned by the `ripserr::vietoris_rips` function.
#' 
#' Note that the absolute value of this measure of distance is not meaningful
#' without a null distribution or at least another value for relative
#' comparison (e.g. finding most similar pair within a triplet).
#' 
#' @param phom1 3-by-n numeric matrix containing persistent homology for first dataset
#' @param phom2 3-by-n numeric matrix containing persistent homology for second dataset
#' @param limit.num limit comparison to only top `limit.num` features in each dimension
#' @return distance vector (1 element per dimension) between `phom1` and `phom2`
#' @export
phom.dist <- function(phom1, phom2, limit.num = 0) {
  # make sure both matrices have at least some features
  if (nrow(phom1) < 1 | nrow(phom2) < 1) {
    stop("Each homology matrix must have at least one feature.")
  }
  
  # make sure homology matrix format is correct
  if (ncol(phom1) != 3 | ncol(phom2) != 3) {
    stop("Each homology matrix must have exactly three columns.")
  }
  
  # none of the values in the homology matrix should be negative
  if (sum(phom1 < 0) > 0 |
      sum(phom2 < 0) > 0) {
    stop("A homology matrix cannot contain any negative values.")
  }
  
  # make sure limit.num is an integer
  if (!(class(limit.num) %in% c("integer", "numeric"))) {
    stop(limit.num, " must be of class integer or numeric.")
  }
  limit.num <- as.integer(limit.num)
  
  # select only top n features from each dimension if `limit.num` parameter > 0
  if (limit.num > 0) {
    phom1 <- phom.topn(phom1, n = limit.num)
    phom2 <- phom.topn(phom2, n = limit.num)
  }
  
  # calculate maximum feature dimension for each matrix
  max.dim1 <- max(phom1[, 1])
  max.dim2 <- max(phom2[, 1])

  # call function that does actual work
  wass_mat_calc(phom1, phom2, pow.val = 1, dim = max(max.dim1, max.dim2))
}

#####PERMUTATION TEST#####
#' Statistical Inference for Topological Data Analysis
#'
#' Conducts a permutation test for nonparametric statistical inference
#' of persistent homology in topological data analysis.
#'
#' The persistent homology of two point clouds are compared with the
#' Wasserstein metric (where Wasserstein-1 is also known as the Earth
#' Mover's Distance). However, the magnitude of the metric for a single pair
#' of point clouds is meaningless without a reference distribution. This
#' function uses a permutation test (permuting the points between the two
#' clouds) as a nonparametric hypothesis test for statistical inference.
#'
#' For more details on permutation tests for statistical inference in
#' topological data analysis, see Robinson A, Turner K. Hypothesis
#' testing for topological data analysis. J Appl Comput Topology. 2017;
#' 1(2): 241-261.<doi:10.1007/s41468-017-0008-7>
#'
#' @param data1 first dataset
#' @param data2 second dataset
#' @param iterations number of iterations for distribution in permutation test
#' @param exponent parameter `p` that returns Wasserstein-p metric
#' @param update if greater than zero, will print a message every `update` iterations
#' @param ... arguments for `ripserr::vietoris_rips` used for each permutation;
#'   this includes the `format`, `dim`, and `threshold` parameters
#' @return list containing results of permutation test
#' @export
permutation_test <- function(data1, data2, iterations,
                             exponent = 1, update = 0, ...) {
  # doesn't work for distance matrices, only point clouds; set dim parameter
  dim <- 1
  list.version <- list(...)
  if (length(list.version) > 0) {
    if (!is.null(list.version$format)) {
      if (list.version$format != "cloud") {
        stop("Permutation tests only work for point clouds.")
      }
    }
    # set dim parameter from ... that user passed
    if (!is.null(list.version$dim)) {
      dim <- as.integer(list.version$dim)
    }
  }
  
  if (update > 0) {
    cat("Starting function\n")
  }
  
  # make sure both are matrices with same number of columns,
  # sufficient number of rows, and no missing values
  if (!("matrix" %in% class(data1)) |
      !("matrix" %in% class(data2))) {
    stop("Both point clouds must be passed as matrices.")
  }
  if (update > 0) {
    cat("PASSED error check #1\n")
  }
  if (nrow(data1) < 2 | nrow(data2) < 2) {
    stop("Both point clouds must have at least 2 points (rows) each.")
  }
  if (update > 0) {
    cat("PASSED error check #2\n")
  }
  if (ncol(data1) < 2 | ncol(data2) < 2) {
    stop("Both point clouds must be in least 2 dimensions (columns) each.")
  }
  if (update > 0) {
    cat("PASSED error check #3\n")
  }
  if (ncol(data1) != ncol(data2)) {
    stop("Both point clouds must have the same number of dimensions.")
  }
  if (update > 0) {
    cat("PASSED error check #4\n")
  }
  if (sum(is.na(data1)) > 0 | sum(is.na(data2)) > 0) {
    stop("There should be no NAs in the point clouds passed to this function.")
  }
  if (update > 0) {
    cat("PASSED error check #5\n")
  }
  class.data1 <- class(data1[1,1])
  class.data2 <- class(data2[1,1])
  allowed.classes <- c("numeric", "integer")
  if (!(class.data1 %in% allowed.classes) |
      !(class.data2 %in% allowed.classes)) {
    stop("Point clouds must be formatted as matrices filled with integers or numerics.")
  }
  if (update > 0) {
    cat("PASSED error check #6\n")
  }
  if (iterations <= 1) {
    stop("Permutation test must have at least 2 iterations (preferably more).")
  }
  if (update > 0) {
    cat("PASSED error check #7\n")
  }
  if (update > 0) {
    cat("Beginning calculations\n")
  }

  # calculate Wasserstein values for actual point clouds (prior to permuting)
  orig.wass <- wass_cloud_calc(data1, data2, exponent, ...)
  
  if (update > 0) {
    cat("Initial calculation complete; starting iterations\n")
  }

  # calculate Wasserstein values for each permutation
  combo.pts <- rbind(data1, data2)
  wass.values <- matrix(-1, ncol = ncol(data1), nrow = iterations)
  worked <- vapply(X = 1:iterations,
                   FUN.VALUE = logical(1),
                   FUN = function(curr.iter) {
                     # print update message if necessary
                     if (update > 0 & curr.iter %% update == 0) {
                       cat("At iteration #", curr.iter, "\n", sep = "")
                     }
                     
                     # permute the point clouds
                     curr.pts <- combo.pts[sample.int(n = nrow(combo.pts),
                                                      size = nrow(combo.pts),
                                                      replace = FALSE), ]

                     # calculate Wasserstein for permuted clouds
                     curr.wass.calc <- wass_cloud_calc(pts1 = curr.pts[1:nrow(data1), ],
                                                       pts2 = curr.pts[(nrow(data1) + 1):nrow(curr.pts), ],
                                                       pow.val = exponent,
                                                       ...)
                     #print(curr.wass.calc)

                     # store into matrix
                     wass.values[curr.iter, ] <<- curr.wass.calc[1:ncol(wass.values)]

                     # all went okay
                     return(TRUE)
                   })

  # return list of lists; each smaller list corresponds to a single dimension
  # each smaller list has: (1) p-value (not corrected multiple testing)
  #                        (2) all trajectories (to plot as histogram)
  #                        (3) Wasserstein original metric for this dimension
  #                        (4) Title (character string with dimension to avoid mismatch w/ indexes)
  answer <- lapply(X = 0:dim,
                   FUN = function(curr.dim) {
                     curr.ans <- list()
                     curr.ans$dimension <- curr.dim
                     curr.ans$permvals <- wass.values[, curr.dim + 1]
                     curr.ans$wasserstein <- orig.wass[curr.dim + 1]
                     names(curr.ans$wasserstein) <- NULL
                     curr.ans$pvalue <- mean(curr.ans$permvals > curr.ans$wasserstein)

                     return(curr.ans)
                   })
  if (update > 0) {
    cat("Completed calculations\n")
  }
  return(answer)
}
